threadpoolctl.py

```diff
@@ -199,11 +155,29 @@ def __init__(self, controller, *, limits=None, user_api=None):
         self._set_threadpool_limits()
 
     def __enter__(self):
+        if not hasattr(self, "_limits_set"):
+            # Set the limits here if the instance was not created via __init__.
+            self._set_threadpool_limits()
         return self
 
     def __exit__(self, type, value, traceback):
         self.unregister()
 
+    @classmethod
+    def wrap(cls, controller, *, limits=None, user_api=None):
+        """Return an instance of this class that can be used as a decorator"""
+        obj = type(cls.__name__, (ContextDecorator,), dict(cls.__dict__))
+
+        # we need to create an instance without calling __init__ because we want the
+        # limits to be set when calling the decorated function, not when creating the
+        # decorator.
+        new_ctx = obj.__new__(obj)
```

Comment:
Wouldn't it be possible to call the constructor without limits and have a private method to update/set the limits after the object has been instantiated without hacking around with `obj.__new__`?

Reply:
I don't think so because when it's used as a callable we only create the instance, so the limits have to be set in the init.

Reply:
I just thought of an alternative but I'm not sure it's less hacky :)
I'll push it and can see if it's better or not

Reply:
So in this alternative implem, the wrap method of _threadpool_limits returns an instance of a child class of this class.

I thought there would be some kind of reference cycle contradiction, but python :)

Do you prefer this solution ?